govcd/service_account.go

```diff
@@ -0,0 +1,371 @@
+	serviceAccount, err := org.GetServiceAccountById(serviceAccountID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Service account by ID: %s", err)
+	}
+
+	return serviceAccount, nil
+}
+
+// Update updates the modifiable fields of a Service ACcount
+func (sa *ServiceAccount) Update(saConfig *types.ServiceAccount) (*ServiceAccount, error) {
+	client := sa.org.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointServiceAccounts
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	saConfig.ID = sa.ServiceAccount.ID
+	saConfig.Name = sa.ServiceAccount.Name
+	saConfig.Status = sa.ServiceAccount.Status
+	urlRef, err := client.OpenApiBuildEndpoint(endpoint, saConfig.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	returnServiceAccount := &ServiceAccount{
+		ServiceAccount: &types.ServiceAccount{},
+		org:            sa.org,
+	}
+
+	err = client.OpenApiPutItem(apiVersion, urlRef, nil, saConfig, returnServiceAccount.ServiceAccount, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error updating Service Account: %s", err)
+	}
+
+	return returnServiceAccount, nil
+}
+
+// Authorize authorizes a service account and returns a DeviceID and UserCode which will be used while granting
+// the request, and sets the Service Account in `Requested` status
+func (sa *ServiceAccount) Authorize() error {
+	client := sa.org.client
+
+	uuid := extractUuid(sa.ServiceAccount.ID)
+	data := bytes.NewBufferString(
+		fmt.Sprintf("client_id=%s",
+			uuid,
+		))
+
+	resp, err := client.doTokenRequest(sa.ServiceAccount.Org.Name, "device_authorization", "37.0", "application/x-www-form-urlencoded", data)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(body, &sa.authParams)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Grant Grants access to the Service Account and sets it in `Granted` status
+func (sa *ServiceAccount) Grant() error {
+	if sa.authParams == nil {
+		return fmt.Errorf("error: userCode is unset, service account needs to be authorized")
+	}
+
+	client := sa.org.client
+	// This is the only place where this field is used, so a local struct is created
+	type serviceAccountGrant struct {
+		UserCode string `json:"userCode"`
+	}
+
+	userCode := &serviceAccountGrant{
+		UserCode: sa.authParams.UserCode,
+	}
+
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointServiceAccountGrant
+	urlRef, err := client.OpenApiBuildEndpoint(endpoint)
+	if err != nil {
+		return fmt.Errorf("error granting service account: %s", err)
+	}
+
+	tenantContext, err := sa.org.getTenantContext()
+	if err != nil {
+		return fmt.Errorf("error granting service account: %s", err)
+	}
+
+	err = client.OpenApiPostItem("37.0", urlRef, nil, userCode, nil, getTenantContextHeader(tenantContext))
```

Comment:
Similar comment as in other place - using `getOpenApiHighestElevatedVersion` would help you get API version in a centralised way and avoid hardcoding API version

Reply:
Fixed.